pkg/controller/hibernation/azure_actuator.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return err
+	}
+	if len(machines) == 0 {
+		logger.Warning("No machines were found to start")
+		return nil
+	}
+	errs := []error{}
+	for _, machineName := range azureMachineNames(machines) {
+		logger.WithField("machine", machineName).Info("Starting cluster machine")
+		_, err = azureClient.StartVirtualMachine(context.TODO(), clusterDeploymentResourceGroup(cd), machineName)
+		if err != nil {
+			errs = append(errs, err)
+			logger.WithError(err).WithField("machine", machineName).Error("Failed to start machine")
+		}
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+// MachinesRunning will return true if the machines associated with the given
+// ClusterDeployment are in a running state.
+func (a *azureActuator) MachinesRunning(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) (bool, error) {
+	logger = logger.WithField("cloud", "azure")
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return false, err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureNotRunningStates, logger)
+	if err != nil {
+		return false, err
+	}
+	return len(machines) == 0, nil
+}
+
+// MachinesStopped will return true if the machines associated with the given
+// ClusterDeployment are in a stopped state.
+func (a *azureActuator) MachinesStopped(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) (bool, error) {
+	logger = logger.WithField("cloud", "azure")
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return false, err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureNotStoppedStates, logger)
+	if err != nil {
+		return false, err
+	}
+	return len(machines) == 0, nil
+}
+
+type azureMachineLister struct {
+	client        azureclient.Client
+	resourceGroup string
+	states        sets.String
+	done          chan struct{}
+	err           chan error
+	logger        log.FieldLogger
+}
+
+func listAzureMachines(cd *hivev1.ClusterDeployment, azureClient azureclient.Client, states sets.String, logger log.FieldLogger) ([]compute.VirtualMachine, error) {
+	page, err := azureClient.ListAllVirtualMachines(context.TODO(), "true")
+	if err != nil {
+		return nil, err
+	}
+	result := []compute.VirtualMachine{}
+	for page.NotDone() {
+		result = append(result, filterByResourceGroupAndState(page.Values(), clusterDeploymentResourceGroup(cd), states, logger)...)
+		if err = page.Next(); err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+func filterByResourceGroupAndState(machines []compute.VirtualMachine, resourceGroup string, states sets.String, logger log.FieldLogger) []compute.VirtualMachine {
+	result := []compute.VirtualMachine{}
+	for _, vm := range machines {
+		if vm.ID == nil {
+			continue
+		}
+		resource, err := azure.ParseResourceID(*vm.ID)
+		if err != nil {
+			logger.WithError(err).Warningf("Failed to parse resource ID")
+			continue
+		}
+		logger.Infof("Parsed resource: %#v", resource)
```

Comment:
This should be debug level (or removed entirely). It outputs a log entry for every machine in the account every time that the machines are listed.